src/Commands/Plugin/PluginBaseCommand.php

```diff
@@ -0,0 +1,334 @@
+                $latest_version = $this->getLatestVersion($plugin_dir);
+                if ($installed_version < $latest_version) {
+                    exec("cd \"$plugin_dir\" && git checkout {$latest_version}", $messages);
+                } else {
+                    $messages[] = "Already up-to-date.";
+                }
+                break;
+
+            case 'composer':
+                $composer_info = $this->getComposerInfo($plugin);
+                $project = $composer_info['name'];
+                $packagist_url = "https://packagist.org/packages/{$project}";
+                if ($this->isValidUrl($packagist_url)) {
+                    exec("rm -rf \"{$plugin_dir}\" && composer create-project --prefer-source --keep-vcs -n -d {$plugins_dir} {$project}:~1", $messages);
+                } else {
+                    $messages[] = "{$packagist_url} is not a valid Packagist project.";
+                }
+                break;
+
+            case 'archive':
+            default:
+                if ($this->commandExists('curl') && $this->commandExists('tar')) {
+                    $project = 'unknown';
+                    $composer_info = $this->getComposerInfo($plugin);
+                    if (!empty($composer_info)) {
+                        $project = $composer_info['name'];
+                    }
+                    $archive_url = "https://github.com/{$project}/archive/1.x.tar.gz";
+                    if ($this->isValidUrl($archive_url)) {
+                        exec("rm -rf \"{$plugin_dir}\" && curl {$archive_url} -L | tar -C {$plugins_dir} -xvz", $messages);
+                    } else {
+                        $messages[] = "{$archive_url} is not a valid Git project.";
+                    }
+                } else {
+                    $messages[] = "In order to update archive plugin projects, you need to install curl and tar.";
+                }
+        }
+        foreach ($messages as $message) {
+            $this->log()->notice($message);
+        }
+    }
+
+    /**
+     * Get the currently installed plugin version.
+     *
+     * @param string $plugin Path to plugin
+     * @return string Installed plugin version
+     */
+    protected function getInstalledVersion($plugin)
+    {
+        exec("cd \"$plugin\" && git describe --tags", $tags);
+        if (!empty($tags)) {
+            $version = array_pop($tags);
+        } else {
+            exec("cd \"$plugin\" && git rev-parse --abbrev-ref HEAD", $branch);
+            $version = !empty($branch) ? array_pop($branch) : 'unknown';
+        }
+        return $version;
+    }
+
+    /**
+     * Get the latest available plugin version.
+     *
+     * @param string $plugin Path to plugin
+     * @return string Latest plugin version
+     */
+    protected function getLatestVersion($plugin)
+    {
+        exec("cd \"$plugin\" && git fetch --all && git tag -l | sort -r | head -1", $tag);
```

Comment:
This should only consider tags that are (a) semver and (b) the same major version as Terminus' major version.